packages/sdk/src/peer_manager/peer_manager.ts

```diff
@@ -29,104 +72,186 @@ export class PeerManager {
+        results.push(peer);
+      }
+    }
+
+    const lockedPeers = results.filter((p) => this.isPeerLocked(p.id));
+
+    if (lockedPeers.length >= this.numPeersToUse) {
+      return lockedPeers.slice(0, this.numPeersToUse).map((p) => p.id);
+    }
+
+    const notLockedPeers = results.filter((p) => !this.isPeerLocked(p.id));
+
+    results = [...lockedPeers, ...notLockedPeers]
+      .slice(0, this.numPeersToUse)
+      .map((p) => {
+        this.lockPeer(p.id);
+        return p;
+      });
+
+    return results.map((p) => p.id);
   }
 
-  public requestRenew(peerId: PeerId | string): PeerId | undefined {
-    const lockedConnections = this.getLockedConnections();
-    const neededPeers = this.numPeersToUse - lockedConnections.length;
+  public async renewPeer(id: PeerId, params: GetPeersParams): Promise<void> {
+    const connectedPeers = await this.connectionManager.getConnectedPeers();
+    const renewedPeer = connectedPeers.find((p) => p.id.equals(id));
 
-    if (neededPeers === 0) {
+    if (!renewedPeer) {
+      log.warn(`Cannot renew peer:${id}, no connection to the peer.`);
       return;
     }
 
-    const connections = this.getUnlockedConnections()
-      .filter((c) => !c.remotePeer.equals(peerId))
-      .slice(0, neededPeers)
-      .map((c) => this.lockConnection(c))
-      .map((c) => c.remotePeer);
+    this.unlockPeer(renewedPeer.id);
+    await this.getPeers(params);
+  }
+
+  public async isPeerOnPubsub(
+    id: PeerId,
+    pubsubTopic: string
+  ): Promise<boolean> {
+    return this.connectionManager.isPeerOnPubsubTopic(id, pubsubTopic);
+  }
 
-    const newPeerId = connections[0];
+  /**
+   * Used to notify about new Filter peer available.
+   */
+  private async onConnected(event: CustomEvent<PeerId>): Promise<void> {
+    const peerId = event.detail;
+    const peer = (await this.connectionManager.getConnectedPeers()).filter(
+      (p) => p.id.equals(peerId)
+    )[0];
 
-    if (!newPeerId) {
+    if (!peer) {
       log.warn(
-        `requestRenew: Couldn't renew peer ${peerId.toString()} - no peers.`
+        `Received connected event for peer:${peerId}, but no connection found.`
       );
       return;
     }
 
-    log.info(
-      `requestRenew: Renewed peer ${peerId.toString()} to ${newPeerId.toString()}`
-    );
+    const isFilterPeer = this.hasPeerProtocol(peer, Protocols.Filter);
+    if (isFilterPeer) {
+      this.dispatchFilterPeerConnect(peer.id);
+    }
+  }
+
+  /**
+   * Used to notify about Filter peer disconnect.
+   */
+  private async onDisconnected(event: CustomEvent<PeerId>): Promise<void> {
+    const peerId = event.detail;
+    let peer = (await this.connectionManager.getConnectedPeers()).filter((p) =>
+      p.id.equals(peerId)
+    )[0];
 
-    return newPeerId;
+    if (peer) {
+      log.warn(
+        `Received disconnected event for peer:${peerId}, but connection is still present.`
+      );
+      return;
+    }
```

Comment:
DRY all this and above logic is duplicated in `onConnected`

Reply:
:) there is a limit to DRY
here I see only detaching condition logic

what's your proposition?